# Path of Exile game stats data eda

dates: 2022.04.20 ~ 

This file contains an analysis processing of the game 'Path of Exile' data.

###### Origins of datas from here: 
poe_stats : https://www.kaggle.com/datasets/gagazet/path-of-exile-league-statistic

Data contains stats of 59000 players, from 4th August of 2017 and before now.

---

This file made for pre-processing the data set for the better analyzing it.

Before the pre-processing, there are some EDA.

판다스 복습과 데이터 전처리에 초점을 둔 파일이며, 그전에 게임을 파악하기 위해 일부 EDA가 함께 있습니다.

# Dataset exploring

## data and libraries import

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

poe_stats = pd.read_csv('poe_stats.csv')

In [2]:
poe_stats

rank   dead  online                    name  level        class  \
0          1  False   False      Tzn_NecroIsFineNow    100  Necromancer   
1          1  False   False        RaizNeverFirstQT    100  Necromancer   
2          1  False   False  GucciStreamerAdvantage    100  Necromancer   
3          1  False   False             ChiroxPrime    100       Slayer   
4          2  False   False     Cool_NecroIsFineNow    100      Deadeye   
...      ...    ...     ...                     ...    ...          ...   
59771  14999   True   False       ПроклятьеРекласта     89  Necromancer   
59772  15000  False   False       IshibashiSummoner     94  Necromancer   
59773  15000  False   False                 BLively     73       Slayer   
59774  15000  False   False                vawddvaw     89    Gladiator   
59775  15000  False   False                 Reselin     53  Necromancer   

                                                      id  experience  \
0      3dcddd59f5088893f734f39686350990dae168cc4f4b00...  4250334444   
1      8f3216db5ac9106c287a834731aafc83c387138f28fc45...  4250334444   
2      c6ec2dae3855c551e0597c06ef2da06fbb5512487ded7f...  4250334444   
3      c861372da792be0b22c45bf437ccd58437c52e9455e0bf...  4250334444   
4      24ae924ceed7989ef3d3d6772612832bb467a6094352c7...  4250334444   
...                                                  ...         ...   
59771  d33b4f6e08c10e365765f9a36a8f36d561fd1d86f10ee1...  1786695653   
59772  5764cfa387e0a87a4bebc1a3c5017e92de8bbb06445c47...  2720770853   
59773  9ac75ab75a47cee8a9dfb0a31912df8909720a8b20f9f1...   495034029   
59774  cf02dfc0c90b2df9c7ac76bbedd91e93c2a8a2ca629c15...  1786638972   
59775  f7ffda5ca2490546344d32930693f8299930ce9451df4a...    77909816   

            account  challenges        twitch              ladder  
0            TheTzn          40        thetzn           Harbinger  
1            RaizQT          22        raizqt    SSF Harbinger HC  
2       GucciPradas          35   guccipradas  Hardcore Harbinger  
3          Chiroxun          26           NaN       SSF Harbinger  
4          cooltail          22           NaN           Harbinger  
...             ...         ...           ...                 ...  
59771  Romanchik002          18  romanchik002  Hardcore Harbinger  
59772  Ishibashisan          33           NaN           Harbinger  
59773        ruby01          19           NaN       SSF Harbinger  
59774        Lecarn          24        lecarn  Hardcore Harbinger  
59775      woworpoe          19           NaN    SSF Harbinger HC  

[59776 rows x 12 columns]

In [3]:
df = poe_stats
df.sample(5)

rank   dead  online            name  level         class  \
17101   4286  False   False        Archessa     87     Ascendant   
15330   3842  False   False         Pilazza     88  Elementalist   
47681  11962   True   False   Raspo_SSF_SRS     60   Necromancer   
19977   5008  False   False  VenomSpinToWin     96      Champion   
13229   3315   True   False     ShaperBlock     93     Gladiator   

                                                      id  experience  \
17101  8954f65c169b5ac29c88de1b0945c4eb9dad57b60027b4...  1569040152   
15330  60927ce9755a037022056b880f83594733fe33d602399d...  1650044959   
47681  b29fc1bd9b7e154ed02055cf1cd9b45e0d1f578e168ae5...   154736353   
19977  0aec8993cef0cb5bbd55bdec29dd319c3a219dae918d7f...  3117099988   
13229  4f4089e809812f51e02f641b1e21be2f80195454b981f5...  2509901064   

          account  challenges twitch              ladder  
17101   archanter          18    NaN       SSF Harbinger  
15330     yrriban          25    NaN       SSF Harbinger  
47681     Raspo42           7    NaN    SSF Harbinger HC  
19977  senzor1996          27    NaN           Harbinger  
13229     LordLan          25    NaN  Hardcore Harbinger

In [4]:
df.shape

(59776, 12)

## data describes

In [5]:
df.columns

Index(['rank', 'dead', 'online', 'name', 'level', 'class', 'id', 'experience',
       'account', 'challenges', 'twitch', 'ladder'],
      dtype='object')

In [6]:
df.describe()

rank         level    experience    challenges
count  59776.000000  59776.000000  5.977600e+04  59776.000000
mean    7495.369379     85.422578  1.825786e+09     25.759569
std     4330.179898     11.288017  1.094748e+09      8.440888
min        1.000000     53.000000  7.790982e+07      4.000000
25%     3745.000000     78.000000  7.774697e+08     19.000000
50%     7492.000000     90.000000  1.934010e+09     25.000000
75%    11247.000000     95.000000  2.876117e+09     33.000000
max    15000.000000    100.000000  4.250334e+09     40.000000

In [7]:
df.describe(include='object')

name      class                                                 id  \
count    59776      59776                                              59776   
unique   58286         26                                              59776   
top     _Flic_  Berserker  3dcddd59f5088893f734f39686350990dae168cc4f4b00...   
freq        14       8711                                                  1   

        account            twitch         ladder  
count     59776             16300          59776  
unique    37123              9613              4  
top     chuz101  thespunkylobster  SSF Harbinger  
freq         30                18          14981

In [8]:
df.nunique()

rank          15000
dead              2
online            2
name          58286
level            48
class            26
id            59776
experience    54124
account       37123
challenges       37
twitch         9613
ladder            4
dtype: int64

In [9]:
df.isnull().sum()

rank              0
dead              0
online            0
name              0
level             0
class             0
id                0
experience        0
account           0
challenges        0
twitch        43476
ladder            0
dtype: int64

### Only column 'twitch' has null
아이디가 없는 사람도, 혹은 한 twitch 아이디에 여러 계정을 가지고 있는 사람도 있을 것이다. 

In [10]:
df['twitch'].describe()

count                16300
unique                9613
top       thespunkylobster
freq                    18
Name: twitch, dtype: object

#### Some twitch users have the plural POE ID 

In [11]:
df['twitch'].groupby(df['twitch']).count().sort_values()

twitch
001100010010         1
newromancers03       1
newfy1991            1
newbieroadkill       1
newancestry          1
                    ..
yoshixt             15
zen1703             15
iispasov            16
seyser_koze         16
thespunkylobster    18
Name: twitch, Length: 9613, dtype: int64

### describe results

데이터 총 개수 59776 개

컬럼명: ['rank', 'dead', 'online', 'name', 'level', 'class', 'id', 'experience', 'account', 'challenges', 'twitch', 'ladder']

랭크는 총 1만 5천개, 죽음과 온라인 여부는 True, False로 이루어져있다.

id가 오직 unique 하며, 한 아이디당 여러 account를 가질 수 있는 것으로 보인다. 그에 따라 일부 name은 겹치는 것을 보인다.

level 과 class는 게임 내 레벨과 직업으로 이루어져 있다.

experience 는 게임 내 경험치를 의미하는 것일 것이며,

twitch는 트위치 계정이 적혀있는 컬럼으로 보인다. 일부 살펴본 결과 한 트위치 아이디로 여러 id를 운영하며 스트리밍 하는 것으로 보인다. 최대 18 아이디를 운영하는 twitch id도 있다.

challenges 의 경우 poe 게임이 던전형 rpg임을 감안하여 짐작해보건대 해당 랭크를 따기까지 해당 모드를 몇 번 플레이 했는지를 나타내는 것으로 예상된다.

##### ladder의 경우 

고유 값이 4개로, 원본 데이터 문서를 찾아본 결과 

```
One file with 12 data sections. One league - Harbinger, but 4 different types of divisions:
Harbinger
Hardcore Harbinger
SSF Harbinger
SSF Harbinger HC
Each division has own ladder with leaders
```
라고 설정되어있다. 

https://pathofexile.fandom.com/wiki/Harbinger_league#Challenges_and_Rewards 와 
https://pathofexile.fandom.com/wiki/Solo_Self-Found 를 참조한 결과 
> Harbinger라는 이름의 시즌 리그의 일종으로 보이며 HC는 Hardcore의 약어로 보인다. Hardcore Harbinger에서 죽은 경우 일반 리그로 옮겨진다는 것과, SSF 가 Solo self-found 로 솔로플레이 모드(파티 거부)의 일종인 점 등을 미루어 보아 난이도는 Harbinger < Hardcore Harbinger < SSF Harbinger < SSF Harbinger HC 가 아닐까 예상된다. 물론 분석에서는 파티와 솔로 플레이 모드는 구분을 해야하며, 난이도 별로도 나눌 수 있을 것이다. 클래스 별 비율 등도 각 게임 모드마다 차이가 날 것이다.

> 또한, 해당 게임이 서버가 세계 섹터 별로 나뉘어있는지는 몰라도 세계적이든 한 지역이든 이만큼의 유저들은 랭커라고 불리울 수 있는 숫자 안에 들은 것이므로 이 분석 자체가 랭커들에 대한 분석이 될 것임을 생각해 볼 수 있다.

> 추가로, SSF가 아닌 일반 모드는 함께 플레이 하면서 랭크를 올릴 수 있기 때문에 다양한 전투 보조 직업이 랭크에 올라갈 수 있으나 SSF 모드의 경우 혼자 딜하고 보조하고 '북치고 장구치고'를 전부 해야하기 때문에 일반적인 보조 직업으로는 상위 랭크를 달성하기 어려울 것으로 예상된다.  

### EDA planning

이에 따라 현 데이터 셋은 크게 이렇게 나누어 볼 수 있을 것이다.

1. rank (rank bins) 
2. level or exp bins
3. class or stats
4. ladder

또한 위의 내용을 laddar 로 나누어 다시 나누어 볼 수 있을 것이다.

# meaning of 'rank' is...? - it has to be considered with 'ladder'

현재 데이터에는 총 15000개의 랭크 고유 개수가 있는데, 데이터는 약 6만개에 이른다.

In [12]:
df['rank'].value_counts().nunique()

3

각 랭크는 최대 3명으로 되어있다. 이것만 보아도 게임 모드에 따라 랭크가 구분되어 있음을 미루어 짐작해 볼 수 있다. 확인해본다.

In [13]:
# 각 모드의 id, rank 고유 개수

modes = ['Harbinger', 'Hardcore Harbinger', 'SSF Harbinger', 'SSF Harbinger HC']

for i in modes:
    print("ladder: ", i)
    print("values of unique id", df[df['ladder'] == i]['id'].nunique())
    print("values of unique rank", df[df['ladder'] == i]['rank'].nunique())
    print('-------------------------------')

ladder:  Harbinger
values of unique id 14918
values of unique rank 14918
-------------------------------
ladder:  Hardcore Harbinger
values of unique id 14905
values of unique rank 14905
-------------------------------
ladder:  SSF Harbinger
values of unique id 14981
values of unique rank 14981
-------------------------------
ladder:  SSF Harbinger HC
values of unique id 14972
values of unique rank 14972
-------------------------------


In [14]:
pd.pivot_table(data = df, index = ['ladder'], aggfunc = 'count')

account  challenges  class   dead  experience     id  \
ladder                                                                     
Harbinger             14918       14918  14918  14918       14918  14918   
Hardcore Harbinger    14905       14905  14905  14905       14905  14905   
SSF Harbinger         14981       14981  14981  14981       14981  14981   
SSF Harbinger HC      14972       14972  14972  14972       14972  14972   

                    level   name  online   rank  twitch  
ladder                                                   
Harbinger           14918  14918   14918  14918    4464  
Hardcore Harbinger  14905  14905   14905  14905    5452  
SSF Harbinger       14981  14981   14981  14981    2661  
SSF Harbinger HC    14972  14972   14972  14972    3723

아이디의 개수와 랭크의 개수가 일치하는 것으로 나왔다. 게임 모드에 따라 랭크가 구분되어 있음이 맞다.

개인적인 생각으로, 이 데이터는 rank별로 나열될 것이 아니라, ladder 별로 rank가 나열되어야 하는 것이 맞다고 생각이 된다.

여기서 한가지 더 해보자면 twitch 아이디가 있는 것을 따져볼 수도 있을 것이다.

## with ladder = df_byladder

In [15]:
df_byladder = df[['ladder', 'rank', 'dead', 'online', 'name', 'level', 'class', 'id', 'experience', 'account', 'challenges', 'twitch']].sort_values(by = ['ladder', 'rank'])
df_byladder.head(5)

ladder  rank   dead  online                  name  level        class  \
0   Harbinger     1  False   False    Tzn_NecroIsFineNow    100  Necromancer   
4   Harbinger     2  False   False   Cool_NecroIsFineNow    100      Deadeye   
8   Harbinger     3  False   False  Rami_The_Negromaniac    100  Necromancer   
13  Harbinger     4  False   False  Guraj_NecroIsFineNow    100       Raider   
18  Harbinger     5  False   False                 Ryphe    100    Occultist   

                                                   id  experience   account  \
0   3dcddd59f5088893f734f39686350990dae168cc4f4b00...  4250334444    TheTzn   
4   24ae924ceed7989ef3d3d6772612832bb467a6094352c7...  4250334444  cooltail   
8   00c0030fc548cc4c0d804564f8fb356610044b7e6d484f...  4250334444  rami1337   
13  34bedf0408f53151dcc31eb43c3bd10a9fb0d7ba32b95b...  4250334444   Guraj93   
18  b6387a6be18939980798db2ec8e8a2924027fa803e063a...  4250334444     Ryphe   

    challenges      twitch  
0           40      thetzn  
4           22         NaN  
8           40    rami1337  
13          24  guraj_hots  
18          40  rypherious

# what 'class'es in there?

In [16]:
df_byladder['class'].unique()

array(['Necromancer', 'Deadeye', 'Raider', 'Occultist', 'Pathfinder',
       'Guardian', 'Champion', 'Gladiator', 'Inquisitor', 'Elementalist',
       'Slayer', 'Berserker', 'Juggernaut', 'Ascendant', 'Trickster',
       'Hierophant', 'Chieftain', 'Saboteur', 'Assassin', 'Ranger',
       'Marauder', 'Scion', 'Witch', 'Duelist', 'Shadow', 'Templar'],
      dtype=object)

In [17]:
df_byladder.groupby(by = ['class']).mean()

rank      dead    online      level    experience  \
class                                                                     
Ascendant      7540.713639  0.367608  0.013319  85.323122  1.785989e+09   
Assassin       8846.387334  0.331370  0.013255  80.048601  1.297246e+09   
Berserker      7601.758122  0.300884  0.013431  86.998967  1.957216e+09   
Champion       7582.100329  0.418586  0.009046  85.018914  1.690701e+09   
Chieftain      7527.660094  0.428638  0.010329  84.022535  1.633104e+09   
Deadeye        8560.173853  0.374131  0.008345  81.350487  1.411977e+09   
Duelist       12337.400000  0.753846  0.015385  61.876923  2.532442e+08   
Elementalist   8517.593514  0.333333  0.012066  80.279789  1.321416e+09   
Gladiator      7086.548367  0.427959  0.010204  84.173673  1.653267e+09   
Guardian       6277.722175  0.380629  0.015293  91.186916  2.485099e+09   
Hierophant     8034.089855  0.437681  0.003478  80.514783  1.278125e+09   
Inquisitor     7826.754940  0.385137  0.007962  83.169272  1.577004e+09   
Juggernaut     7345.391626  0.433990  0.008374  83.238424  1.552577e+09   
Marauder      10884.610526  0.631579  0.010526  64.778947  3.539756e+08   
Necromancer    7162.736329  0.383213  0.015402  84.712449  1.739417e+09   
Occultist      7617.948052  0.424513  0.004058  82.319805  1.508447e+09   
Pathfinder     6856.042935  0.087992  0.040576  94.060392  2.924074e+09   
Raider         7667.577864  0.236117  0.024660  87.094951  2.023136e+09   
Ranger        11337.166667  0.458333  0.000000  66.979167  5.301023e+08   
Saboteur       7714.562842  0.377049  0.004684  82.901639  1.545837e+09   
Scion         11941.596154  0.615385  0.000000  63.557692  3.042050e+08   
Shadow        12328.434783  0.608696  0.000000  62.869565  2.822425e+08   
Slayer         7366.899909  0.332726  0.014221  87.201641  1.983167e+09   
Templar       12541.802817  0.732394  0.000000  60.366197  1.928275e+08   
Trickster      7378.675278  0.419228  0.007521  82.643558  1.464714e+09   
Witch         12136.067227  0.655462  0.000000  62.781513  2.855004e+08   

              challenges  
class                     
Ascendant      26.141449  
Assassin       23.066274  
Berserker      26.481690  
Champion       24.095395  
Chieftain      24.358685  
Deadeye        23.168289  
Duelist        15.676923  
Elementalist   22.068627  
Gladiator      23.921633  
Guardian       30.151232  
Hierophant     21.726377  
Inquisitor     24.452374  
Juggernaut     23.739901  
Marauder       16.021053  
Necromancer    24.990957  
Occultist      24.951299  
Pathfinder     32.499646  
Raider         27.277282  
Ranger         17.645833  
Saboteur       25.249024  
Scion          14.961538  
Shadow         14.847826  
Slayer         27.014950  
Templar        13.140845  
Trickster      23.881949  
Witch          14.697479

## Kinds of Path of Exile classes

해당 데이터셋에서, poe에는 ['Necromancer', 'Deadeye', 'Raider', 'Occultist', 'Pathfinder',
       'Guardian', 'Champion', 'Gladiator', 'Inquisitor', 'Elementalist',
       'Slayer', 'Berserker', 'Juggernaut', 'Ascendant', 'Trickster',
       'Hierophant', 'Chieftain', 'Saboteur', 'Assassin', 'Ranger',
       'Marauder', 'Scion', 'Witch', 'Duelist', 'Shadow', 'Templar']  총 24개의 unique 클래스가 존재하며,

POE 공식 사이트를 참고한 결과 각 직업별 클래스는 이렇게 나뉘고 있다.

    Duelist - 'Slayer', 'Gladiator', 'Champion'
    Shadow - 'Assassin', 'Saboteur', 'Trickster'
    Marauder(약탈자 라고 한다) - 'Juggernaut', 'Berserker', 'Chieftain',
    Witch - 'Necromancer', 'Elementalist', 'Occultist'
    Ranger - 'Raider', 'Deadeye', 'Pathfinder'
    Templar - 'Inquisitor', 'Hierophant', 'Guardian'
    Scion - 'Ascendant'
    
일반적으로 레인저는 원딜러, 듀얼리스트는 전사로 보이며, 위치는 마딜러, 템플러는 수도승 같은 등등 다른 게임에서 보이는 직업과 비슷한 부분이 있지만,

각 클래스가 가진 특성들도 다르기 때문에 poe에서 제공하고있는 클래스 별로도 나눌 수 있을 것이고, 일반적인 rpg 특성에 따라 이를 나누는 방법은 이런 방식들이 될 것이다.

    1. 탱커, 딜러, 힐러(보조직업)
    2. 탱커, 일반딜러, 암살딜러, 힐러(보조직업)
    2. 탱커, 근접딜러, 원거리딜러, 힐러(보조직업)
    3. 탱커, 근접딜러, 원거리물리딜러, 원거리마법딜러, 힐러(보조직업)
    4. 근거리, 원거리, 마법, 힐러(보조직업)
    5. 점사딜스킬위주, 광역딜스킬위주, 힐러(보조직업) ... 등등

이에 따라 각 클래스를 읽어보고 데이터 labeling을 진행하도록 한다.

스킬 범위나 세부 효과에 관련한 부분은 실제로 poe를 해본 사람이 분석하는 것이 더 바를 것 같다는 판단이 든다.

클래스의 호불호 혹은 성능을 떠나서, 세부적인 수치와 관련해서는 플레이 해보지 않는 이상 체감을 해볼 수 없기 때문이다.

예를 들어 템플러의 가디언은 방어 버프 혹은 적의 버프를 방해하는 스킬들이 있는데 상세히 알기는 어렵다.

---

https://hong-sam.tistory.com/27 글을 참조하여 일반적인 직업으로 나누는 것을 먼저 하기로 결정.

글에 따르면 
```
    Duelist : 빠른 근접 전투(힘, 민첩)
    Shadow : 민첩하고 지능적인 전투(민첩, 지능)
    Marauder : 전사(힘)
    Witch : 마법사(지능)
    Ranger : 궁수(민첩)
    Templar : 성기사(힘, 지능)
    Scion : 모든 스탯이 중간(힘, 민첩, 지능)
```

으로 나누어져있다. 사이온 같은 경우 성장 중 사이온을 구하는 퀘스트를 깨야 직업을 할 수 있다고 한다.

스킬을 찍는 데에 자유로운 것이 특징이나 자신의 원래 직업과 스탯에 따라 가는 것이 더 좋은 효율을 낼 수 있다는 점에서 다른 게임과 크게 다르지는 않다.

살펴본 결과 패시브 혹은 버프(디버프)스킬들도 많이 포함되어 있었다. 

자유도가 높지만 높아도 그 안에서 더 좋은 효율을 내는 일부 직업이나 스킬트리가 유행하여 해당 직업에는 한 클래스가 주를 이루기 마련인데, 그런 경향도 있는지 살펴보아야 겠다.

문제는 사이온인데, 사람마다 찍는 유형에 따라 근접일 수도 원거리일 수도, 완전히 보조 일수도, 그 중간 어딘가일 수도 있다.

여러 능력의 조합을 사용하는 것의 이점과 단점은 극명하기 때문에 숙련된 사람이 아니면 사이온을 달고 랭킹에 진입하기는 쉽지 않았을 것이라는 생각도 든다.


방식 1: 세부 클래스 전 직업(전직 전)으로 묶은 다음 그에 따라 labeling, 사이온은 사이온으로.
```
    근 - Duelist - 'Slayer', 'Gladiator', 'Champion'
    근 - Shadow - 'Assassin', 'Saboteur', 'Trickster'
    근 - Marauder - 'Juggernaut', 'Berserker', 'Chieftain',
    원 - Witch - 'Necromancer', 'Elementalist', 'Occultist'
    원 - Ranger - 'Raider', 'Deadeye', 'Pathfinder'
    원 - Templar - 'Inquisitor', 'Hierophant', 'Guardian'
    사이온 - Scion - 'Ascendant'
```
방식 2: 세부 클래스(전직 후)에 따라 labeling (딜 위주, 탱킹 위주, 보조 위주 - 게임 공부를 더 하고 나누는 걸로. EDA 하다보면 아이디어를 얻을수도 있고,)

    'Slayer' : 양손 무기, 생존력 특화 Duelist
    'Gladiator' : 쌍수 혹은 한손-방패 Duelist
    'Champion' : 적 디버프, 자신 버프 Duelist
    'Assassin' : 출혈, 독 등 이용한 치명 딜 위주 Shadow
    'Saboteur' : 트랩, 마인 활용 Shadow
    'Trickster' : 기동성, 회피, 공격속도 위주 Shadow
    'Juggernaut' : 탱킹 중심 Marauder
    'Berserker' : 흡혈이나 피해 비례 데미지 등 광전사 Marauder
    'Chieftain' : 화염, 적 디버프 이용 Marauder
    'Necromancer' : 소환술 특화 Witch
    'Elementalist' : 원소 피해 특화 Witch
    'Occultist' : 카오스 피해, 저주 특화 Witch
    'Raider' : 회피, 처치 후 얻는 버프를 이용하는 Ranger
    'Deadeye' : 치명타, 딜 중심 Ranger
    'Pathfinder' : 플라스크(물약), 치유효과 강화, 디버프 제거, 원소, 카오스 이용 Ranger
    'Inquisitor' : 원소 치명타, 마나 재생, 자신 버프 Templer
    'Hierophant' : 토템 및 마나 변환, 받은 피해 마나로 감소 등의 Templer
    'Guardian' : 서포트 및 보호(방어버프, 디버프제거, 치유 등) 특화 Templer
    'Ascendant' : 타 클래스들의 조합하는 Scion. 성능은 다소 떨어질 수 있음.

방식 3: 특정 스탯 한 가지 / 혹은 두 가지 / 세 가지

    'Marauder', 'Witch', 'Ranger'
    'Duelist', 'Shadow', 'Templar'
    'Scion'

In [18]:
df_byladder['class'].unique()

array(['Necromancer', 'Deadeye', 'Raider', 'Occultist', 'Pathfinder',
       'Guardian', 'Champion', 'Gladiator', 'Inquisitor', 'Elementalist',
       'Slayer', 'Berserker', 'Juggernaut', 'Ascendant', 'Trickster',
       'Hierophant', 'Chieftain', 'Saboteur', 'Assassin', 'Ranger',
       'Marauder', 'Scion', 'Witch', 'Duelist', 'Shadow', 'Templar'],
      dtype=object)

전직 전과 전직 후 *모두 섞여있는 것* 또한 볼 수 있다. 이에 따라 이를 기준으로 나눌 수도 있다. 
전직 전과 후로 찍을 수 있는 스킬이 다를 것이기 때문이다. (물론 젬이 필요하다고 했기 때문에 그것이 없어서 그 직업을 운용중일 지도 모른다)

# class labeling

## ['special_class']: before and after to select special classes

In [19]:
df

rank   dead  online                    name  level        class  \
0          1  False   False      Tzn_NecroIsFineNow    100  Necromancer   
1          1  False   False        RaizNeverFirstQT    100  Necromancer   
2          1  False   False  GucciStreamerAdvantage    100  Necromancer   
3          1  False   False             ChiroxPrime    100       Slayer   
4          2  False   False     Cool_NecroIsFineNow    100      Deadeye   
...      ...    ...     ...                     ...    ...          ...   
59771  14999   True   False       ПроклятьеРекласта     89  Necromancer   
59772  15000  False   False       IshibashiSummoner     94  Necromancer   
59773  15000  False   False                 BLively     73       Slayer   
59774  15000  False   False                vawddvaw     89    Gladiator   
59775  15000  False   False                 Reselin     53  Necromancer   

                                                      id  experience  \
0      3dcddd59f5088893f734f39686350990dae168cc4f4b00...  4250334444   
1      8f3216db5ac9106c287a834731aafc83c387138f28fc45...  4250334444   
2      c6ec2dae3855c551e0597c06ef2da06fbb5512487ded7f...  4250334444   
3      c861372da792be0b22c45bf437ccd58437c52e9455e0bf...  4250334444   
4      24ae924ceed7989ef3d3d6772612832bb467a6094352c7...  4250334444   
...                                                  ...         ...   
59771  d33b4f6e08c10e365765f9a36a8f36d561fd1d86f10ee1...  1786695653   
59772  5764cfa387e0a87a4bebc1a3c5017e92de8bbb06445c47...  2720770853   
59773  9ac75ab75a47cee8a9dfb0a31912df8909720a8b20f9f1...   495034029   
59774  cf02dfc0c90b2df9c7ac76bbedd91e93c2a8a2ca629c15...  1786638972   
59775  f7ffda5ca2490546344d32930693f8299930ce9451df4a...    77909816   

            account  challenges        twitch              ladder  
0            TheTzn          40        thetzn           Harbinger  
1            RaizQT          22        raizqt    SSF Harbinger HC  
2       GucciPradas          35   guccipradas  Hardcore Harbinger  
3          Chiroxun          26           NaN       SSF Harbinger  
4          cooltail          22           NaN           Harbinger  
...             ...         ...           ...                 ...  
59771  Romanchik002          18  romanchik002  Hardcore Harbinger  
59772  Ishibashisan          33           NaN           Harbinger  
59773        ruby01          19           NaN       SSF Harbinger  
59774        Lecarn          24        lecarn  Hardcore Harbinger  
59775      woworpoe          19           NaN    SSF Harbinger HC  

[59776 rows x 12 columns]

In [20]:
df['special_class'] = 0

In [21]:
pre_classes = ['Duelist', 'Shadow', 'Marauder', 'Witch', 'Ranger', 'Templar', 'Scion']

In [22]:
df['special_class'] = np.where(df['class'].isin(pre_classes), 0, 1)

In [23]:
df.head(10)

rank   dead  online                    name  level        class  \
0     1  False   False      Tzn_NecroIsFineNow    100  Necromancer   
1     1  False   False        RaizNeverFirstQT    100  Necromancer   
2     1  False   False  GucciStreamerAdvantage    100  Necromancer   
3     1  False   False             ChiroxPrime    100       Slayer   
4     2  False   False     Cool_NecroIsFineNow    100      Deadeye   
5     2  False   False  SteelStreamerAdvantage    100      Deadeye   
6     2  False   False   Grom_WhyGodThreeCmWhy    100    Gladiator   
7     2  False   False          beobio_capitan    100   Inquisitor   
8     3  False   False    Rami_The_Negromaniac    100  Necromancer   
9     3  False   False           FutonBlewAway    100       Raider   

                                                  id  experience  \
0  3dcddd59f5088893f734f39686350990dae168cc4f4b00...  4250334444   
1  8f3216db5ac9106c287a834731aafc83c387138f28fc45...  4250334444   
2  c6ec2dae3855c551e0597c06ef2da06fbb5512487ded7f...  4250334444   
3  c861372da792be0b22c45bf437ccd58437c52e9455e0bf...  4250334444   
4  24ae924ceed7989ef3d3d6772612832bb467a6094352c7...  4250334444   
5  51acb8e6fac3cfaa7ff848151dad01b0a6f350e7525543...  4250334444   
6  5d8b80b2bdd30f3dba1036a6791187562758935e8d4723...  4250334444   
7  65c58f9b186e1fbe9f927484f3f178cdb9ee0dc2427289...  4250334444   
8  00c0030fc548cc4c0d804564f8fb356610044b7e6d484f...  4250334444   
9  46e56572ff6e403f45fac597240e13e9af225f06ca4e52...  4250334444   

         account  challenges        twitch              ladder  special_class  
0         TheTzn          40        thetzn           Harbinger              1  
1         RaizQT          22        raizqt    SSF Harbinger HC              1  
2    GucciPradas          35   guccipradas  Hardcore Harbinger              1  
3       Chiroxun          26           NaN       SSF Harbinger              1  
4       cooltail          22           NaN           Harbinger              1  
5      Steelmage          37  steelmage202  Hardcore Harbinger              1  
6     rocklegend          22       gromick    SSF Harbinger HC              1  
7  TishaVosemsot          40        beobio       SSF Harbinger              1  
8       rami1337          40      rami1337           Harbinger              1  
9        pleaGRC          40           NaN       SSF Harbinger              1

### How many ranked users in there who didn't select the special class?

In [24]:
df['special_class'].value_counts()

1    59280
0      496
Name: special_class, dtype: int64

In [25]:
# 위와 같은 결과를 그룹바이로 낼 수 있다.

freq = df.groupby(['special_class']).size() 
print(freq)

special_class
0      496
1    59280
dtype: int64


### Then, what pre-classes of them are there?

In [26]:
df[df['special_class'] == 0]['class'].value_counts()

Witch       119
Marauder     95
Templar      71
Duelist      65
Scion        52
Ranger       48
Shadow       46
Name: class, dtype: int64

위 결과를 보니 더더욱! 라벨링을 하는게 좋겠다는 생각이 드는데, 게임을 어떤 관점에서 볼지에 따라 나누는 방식을 달리할 수 있을 것이다.

포지션 자체인지 아니면 전체적인 스킬의 유형(사거리 등)에 관한 것인지 사람마다 중요시 하는 부분이 다르기 때문에... 

살펴보면 살펴볼 수록 이 게임은 사거리 자체보다는 포지션이나, 솔플 난이도 여부에 따라 갈리는게 나을 것 같다는 생각이 든다. 그래서 SSF 모드가 따로 있는 것이기도 한 이유일 것이다.

### What play modes of them are there?

In [27]:
df[df['special_class'] == 0]['ladder'].value_counts()

SSF Harbinger HC      412
SSF Harbinger          78
Hardcore Harbinger      6
Name: ladder, dtype: int64

독특하게도, 전직하지 않은 사람들은 일반 Harbinger 에는 없었고, 대부분 솔로 플레이모드에 많았다.

아무래도 전직이라는 건 세부 직업을 고르는 것, 즉 특화를 하는 것이기 때문에 특화되지 않은 다른 면에서는 부족해질 수 있다고 생각해보면, 

솔로플레이에서는 전직하지 않은 직업으로 플레이 하는게 더 나을 수도 있다는 생각이다. 물론 스킬을 찍는 포인트가 한정되어있다는 사실 안에 말이다.



###  df_pre_cl : stats description by whether pre or post special classes

전직을 하지 않은 사람들과 한 사람들의 통계적 요약을 보고자 한다.

In [28]:
df_pre_cl = df[df['special_class'] == 0]
df_pre_cl

rank   dead  online             name  level     class  \
4483    1123   True   False          Venkula     88     Witch   
5230    1310  False   False  FrozenTheocracy     88  Marauder   
9003    2256  False   False       Wintersorg     85  Marauder   
9220    2311  False   False         Czardone     94    Ranger   
9571    2399  False   False      DuDuCia_SSF     90  Marauder   
...      ...    ...     ...              ...    ...       ...   
59635  14965  False   False          Bratakk     53   Templar   
59672  14975   True   False           zyxzcb     53     Scion   
59717  14986   True   False        flackoraz     53  Marauder   
59731  14989   True   False          Aralena     53     Scion   
59746  14993  False   False        CrankyMin     53     Witch   

                                                      id  experience  \
4483   ff2158b95cc2f345f58a749dc5b87479f3c6c869c2c56d...  1742518167   
5230   d8bee7876d1d73ee030c7d2574d0d2270c9446a5a2c786...  1661086034   
9003   7b93ae4d6bba754e6cebc50518b35f9cea736baf0b58e0...  1308994834   
9220   5cd690e7d5b2092c1c763782a9d20440edd4fd6bcac1c5...  2678680422   
9571   dd90b4d7630137bcc29e761c8d5da905054894071461ea...  1934009687   
...                                                  ...         ...   
59635  ab43cb5747cbcfdaeca5e92c4fd9379a8a4035cf564980...    78452021   
59672  130a9fc1363dbc9867d34b7e557e586400c0c8026f6681...    78307968   
59717  8a6eb406ad742208f1f51530be2cb76a5a6b6cf0894f3f...    78083773   
59731  dc4404dae50544f86092ad1755d0b6b0f4e8c5a8a567f7...    78059267   
59746  69e3b0c17b0d75089bbfd52052acb2aed4bb46b630c03d...    77984723   

             account  challenges         twitch              ladder  \
4483          Baacax          25            NaN    SSF Harbinger HC   
5230      Warlock279          24            NaN    SSF Harbinger HC   
9003      Sandman421          14            NaN    SSF Harbinger HC   
9220   azzmasterking          15            NaN  Hardcore Harbinger   
9571     shanghaibf4          21            NaN       SSF Harbinger   
...              ...         ...            ...                 ...   
59635     WizardOfUz           8            NaN    SSF Harbinger HC   
59672   BlindMessiah          17            NaN    SSF Harbinger HC   
59717        Virchan          16            NaN    SSF Harbinger HC   
59731        Imladir          18  imladir_talen    SSF Harbinger HC   
59746     crankyzola           8            NaN    SSF Harbinger HC   

       special_class  
4483               0  
5230               0  
9003               0  
9220               0  
9571               0  
...              ...  
59635              0  
59672              0  
59717              0  
59731              0  
59746              0  

[496 rows x 13 columns]

In [29]:
df_pre_cl.describe()

rank       level    experience  challenges  special_class
count    496.000000  496.000000  4.960000e+02  496.000000          496.0
mean   11900.975806   63.195565  3.064528e+08   15.183468            0.0
std     2798.356468    9.487475  3.707211e+08    6.326839            0.0
min     1123.000000   53.000000  7.798472e+07    5.000000            0.0
25%    10924.750000   55.000000  9.925252e+07   10.000000            0.0
50%    12580.500000   60.000000  1.559193e+08   14.000000            0.0
75%    14080.250000   70.000000  3.837313e+08   18.000000            0.0
max    14993.000000   94.000000  2.678680e+09   37.000000            0.0

평균 랭크 자체는 높은 편이아니고, 레벨의 최대 값은 94였다. 최대 레벨이 100인 점을 감안할 때 물음표가 생길 수 밖에 없는 부분이다. 게임을 정확히 이해하지 않고서는 풀기 어려운 의문인 것 같다.

반대로 전직을 한 사람들을 보자면, 

In [30]:
df[df['special_class'] == 1].describe()

rank         level    experience    challenges  special_class
count  59280.000000  59280.000000  5.928000e+04  59280.000000        59280.0
mean    7458.507355     85.608553  1.838498e+09     25.848060            1.0
std     4321.827856     11.116002  1.089898e+09      8.400402            0.0
min        1.000000     53.000000  7.790982e+07      4.000000            1.0
25%     3716.000000     79.000000  7.992337e+08     20.000000            1.0
50%     7440.000000     90.000000  1.934066e+09     25.000000            1.0
75%    11190.000000     95.000000  2.876117e+09     33.000000            1.0
max    15000.000000    100.000000  4.250334e+09     40.000000            1.0

세부 클래스로 전직한 평균 랭크가 약 3천 이상 더 높고, 레벨도 평균적으로 더 높다. 

하지만 경험치는 그들보다 평균적으로 더 적었고, 시도도 더 많이 한 것으로 나타났다.

이를 토대로 합리적의심을 해보자면, 숙련도나 게임 경력에 따라서 전직 여부를 결정하지는 않았을 것이라고 추측해 볼 수 있다.

## ['stat_Str', 'stat_Int', 'stat_Dex']: columns by kind of stats and stats numbers

먼저, 특정 스탯 한 가지 / 혹은 두 가지 / 세 가지

    'Marauder', 'Witch', 'Ranger'
    'Duelist', 'Shadow', 'Templar' 
    'Scion'

에 따라 stat_num(개수) 을 1, 2, 3으로 나누게 될 것이다.

각 클래스에 따라 어떤 특성을 찍었는지 특성(힘, 민첩, 지능) 을 나누어서 여부를 라벨링 한 다음 sum 으로 낼 계획


In [31]:
# 각 클래스 나누기 (전직하지 않은 사람들도 있기 때문에 각 전직 이전 직업도 넣어준다.)

# 한 개인 직업
Marauder = ['Marauder', 'Juggernaut', 'Berserker', 'Chieftain'] # Strength
Witch = ['Witch', 'Necromancer', 'Elementalist', 'Occultist'] # Intelligence
Ranger = ['Ranger', 'Raider', 'Deadeye', 'Pathfinder'] # Dexterity

# 두 개인 직업
Duelist = ['Duelist', 'Slayer', 'Gladiator', 'Champion'] # Strength, Dexterity
Shadow = ['Shadow', 'Assassin', 'Saboteur', 'Trickster'] # Dexterity, Intelligence
Templar = ['Templar', 'Inquisitor', 'Hierophant', 'Guardian'] # Intelligence, Strength

# 세 개인 직업
Scion = ['Scion', 'Ascendant'] # Strength, Dexterity, Intelligence

In [32]:
# 앞에 stat을 굳이 붙인 이유는 다 쓰긴 길고 자르자니 str, int 이런 문자여서 보기가 애매하여 붙였다.

df['stat_Str'] = 0
df['stat_Int'] = 0
df['stat_Dex'] = 0

In [33]:
# Strength
df['stat_Str'] = np.where(df['class'].isin(['Marauder', 'Juggernaut', 'Berserker', 'Chieftain', 
                                            'Duelist', 'Slayer', 'Gladiator', 'Champion', 
                                            'Templar', 'Inquisitor', 'Hierophant', 'Guardian',
                                            'Scion', 'Ascendant']), 1, 0)

# Intelligence
df['stat_Int'] = np.where(df['class'].isin(['Witch', 'Necromancer', 'Elementalist', 'Occultist', 
                                            'Shadow', 'Assassin', 'Saboteur', 'Trickster',
                                            'Templar', 'Inquisitor', 'Hierophant', 'Guardian',
                                            'Scion', 'Ascendant']), 1, 0)

# Dexterity
df['stat_Dex'] = np.where(df['class'].isin(['Ranger', 'Raider', 'Deadeye', 'Pathfinder',
                                            'Duelist', 'Slayer', 'Gladiator', 'Champion',
                                            'Shadow', 'Assassin', 'Saboteur', 'Trickster',
                                            'Scion', 'Ascendant']), 1, 0)

In [34]:
df.sample(20)

rank   dead  online                    name  level        class  \
3301     827  False   False                Araktura    100       Slayer   
52632  13203  False   False           HEFEWEISSBIER     75    Gladiator   
34995   8773  False   False             Storm_Vault     81   Inquisitor   
13190   3305   True   False           WhoGotDaProps     93   Hierophant   
38005   9529   True   False              TatarDozor     90   Inquisitor   
45287  11361   True   False  TotemRampage_HARBINGER     90    Chieftain   
37884   9499   True   False               Utterotad     68       Slayer   
55347  13888  False   False              Fescioness     74    Ascendant   
10147   2543  False   False                Raglione     89    Berserker   
55438  13911  False   False              whaledaddy     89    Berserker   
56954  14291  False   False                DavisMom     74    Chieftain   
7398    1854   True   False        XelSSFNeverlucky     86       Slayer   
49424  12399  False   False      FukkatuNoDischarge     89   Juggernaut   
53273  13364  False   False              Mattthelow     94  Necromancer   
5726    1435   True   False               BlurriSRF     87   Inquisitor   
36528   9158  False   False               Crimson__     90       Slayer   
45654  11453   True   False           TotempoleTKoN     62    Chieftain   
9770    2449  False   False             ElyxIceShot     97    Ascendant   
34213   8577   True   False  deadeyeportraitbuffplz     69      Deadeye   
26898   6742  False   False         JumaltenMestari     95       Raider   

                                                      id  experience  \
3301   2c113ca602218b0b22b78ba45ed625318e5afa8df78561...  4250334444   
52632  c595a039704b224bc16c63648d2ad44122d27b41816212...   585801035   
34995  23c6283f73d508b9b30abc6e6fd8bede94f78274c938ba...   977136224   
13190  cfc67307ceda258879d4d454bad66179594d912da399b6...  2510809312   
38005  dca00871b50076536fa4a3fc7d109663adf8bd9809aee5...  2012538070   
45287  806251c6bb13220080fafe441b45f69d51048bd83d569a...  1938848245   
37884  a41ada97ff335005e3ad4eb0d02d5a0cdc8df66ca88a4e...   304972236   
55347  5523375571b1606457cb71f684fd79992711326571eecf...   548132323   
10147  8a3c1969a1c32f090e4f17071cef3ff3aae625b0d8eeb9...  1871582671   
55438  042e8098c658d8695719851ecfe8851a58acee469c3d9e...  1826325113   
56954  568b065e492889c9ee7d0893bbbd9041b6553cdf737a8b...   527677523   
7398   59e279561548d3e2dae5f14b1062e4a2592578e1d419df...  1455554307   
49424  77ddc05361472a265deb340d7cc01caeec5c5ff3197c0a...  1896341696   
53273  d24c2185a818ce3c0564f326dbc22df94f02c3c4731287...  2829480475   
5726   59dda1d2526d045e42bbb9ae8fc1ddc38cce69a793e538...  1607227114   
36528  d6fd78ef69d9c642d030d23c3ab8d1df95a261b4484ae8...  2031833493   
45654  90255a3fc398883ccc590cbafc45c0d5742547aae5a2ff...   177312138   
9770   982c2b9d0a052d5581e3f075309f810598c10944c7e2de...  3429930244   
34213  d62b9bc7b5b7f70a16451170adfe30ae46c889a97474a1...   349372342   
26898  032565778dd468d6cbb91ec08ffa77583d342d13f1e5fd...  2985829552   

             account  challenges          twitch              ladder  \
3301       Sinserrar          26       sinserrar           Harbinger   
52632   ClaireVoyant          12             NaN       SSF Harbinger   
34995      Naiesv0rn          25             NaN       SSF Harbinger   
13190   TheLastWound          24             NaN  Hardcore Harbinger   
38005     Lorronssir          24             NaN  Hardcore Harbinger   
45287    taigeahetsu          29             NaN  Hardcore Harbinger   
37884        Chrollo          24       cokelord5    SSF Harbinger HC   
55347  WalkingWraith          14             NaN       SSF Harbinger   
10147  Tomviolence89          25   tomviolence89       SSF Harbinger   
55438           GCRS          19    haha_meymeys  Hardcore Harbinger   
56954   deadlyweasel          14             NaN       SSF Harbinger   
7398         Exelice          31             NaN    SSF

#### checking

In [35]:
# check
# 위에서 만들었던 리스트를 리스트 안에 넣는다.
pre_classes_list = [Marauder, Witch, Ranger, Duelist, Shadow, Templar, Scion]


for i in pre_classes_list:
    print("Class names: ", i)
    print("Class nums: ", df['class'].isin(i).sum())
    print(df[df['class'].isin(i)][['stat_Str','stat_Int', 'stat_Dex']].sum())
    print("================================================================")

Class names:  ['Marauder', 'Juggernaut', 'Berserker', 'Chieftain']
Class nums:  12966
stat_Str    12966
stat_Int        0
stat_Dex        0
dtype: int64
Class names:  ['Witch', 'Necromancer', 'Elementalist', 'Occultist']
Class nums:  9754
stat_Str       0
stat_Int    9754
stat_Dex       0
dtype: int64
Class names:  ['Ranger', 'Raider', 'Deadeye', 'Pathfinder']
Class nums:  10156
stat_Str        0
stat_Int        0
stat_Dex    10156
dtype: int64
Class names:  ['Duelist', 'Slayer', 'Gladiator', 'Champion']
Class nums:  11666
stat_Str    11666
stat_Int        0
stat_Dex    11666
dtype: int64
Class names:  ['Shadow', 'Assassin', 'Saboteur', 'Trickster']
Class nums:  5064
stat_Str       0
stat_Int    5064
stat_Dex    5064
dtype: int64
Class names:  ['Templar', 'Inquisitor', 'Hierophant', 'Guardian']
Class nums:  6364
stat_Str    6364
stat_Int    6364
stat_Dex       0
dtype: int64
Class names:  ['Scion', 'Ascendant']
Class nums:  3806
stat_Str    3806
stat_Int    3806
stat_Dex    3806
dtype:

제대로 들어간 것을 확인했다. 

## ['stats_num']: How many stats they set on main?

다음은 주스탯 컬럼의 여부에 따라서 능력치를 몇 가지를 사용하고 있는지를 따져볼 차례이다.

In [36]:
df.sample(5)

rank   dead  online                name  level        class  \
40681  10202  False   False         MilkmanDead     79    Trickster   
13350   3345  False   False  OtakuFidgetSpinner     88     Champion   
15399   3859  False   False        rainbow_wind     96   Pathfinder   
13947   3495  False   False         HCSSF_Clare     81  Necromancer   
56721  14233  False   False         Pivchanskii     94    Chieftain   

                                                      id  experience  \
40681  3620dbe5e49e9c18e14c3cf797fd227b2de80aad7d2acd...   822772853   
13350  f0de8217f250c404677752b79902b1d91ecb0502d63f55...  1709664774   
15399  672353ea6f2d4e64961416a4c373a2dd4db9fef45c01ec...  3238661733   
13947  ded81a6ce1e122a6cda46bb22a2bace0759027441a35e1...   952559186   
56721  6c3299955176c45857032b604c63b80bbf5f8db1f651e4...  2757737254   

           account  challenges twitch            ladder  special_class  \
40681        dimms          24    NaN     SSF Harbinger              1   
13350       angepo          18    NaN     SSF Harbinger              1   
15399     Bont3006          36    NaN         Harbinger              1   
13947  veelersbirb          29    NaN  SSF Harbinger HC              1   
56721    Pivasik99          34    NaN         Harbinger              1   

       stat_Str  stat_Int  stat_Dex  
40681         0         1         1  
13350         1         0         1  
15399         0         0         1  
13947         0         1         0  
56721         1         0         0

In [37]:
# stats_num is the number of main stats for the class

df['stats_num'] = df['stat_Str'] + df['stat_Int'] + df['stat_Dex']

# checking

df.sample(5)

rank   dead  online                    name  level        class  \
3724     933  False   False               KomeePete     96     Champion   
4406    1104  False   False            kenntanakaHA     91  Necromancer   
36748   9214  False   False  Harbringer_TwoPointTwo     81       Raider   
41356  10371   True   False             SmilingMoon     90   Hierophant   
59087  14827   True   False       Mushiiiiiiiiiiiii     89    Berserker   

                                                      id  experience  \
3724   c0b4392f49121e29119a3d7d21ad87bce528e251aa1cdd...  3111558534   
4406   21bc30ef7aa56a6415e486674d82df44799957d7bf0881...  2152504548   
36748  049f16da79fce345af8450bcd5720014b9fc655b5e794a...   930580995   
41356  f810d23c1ff98976b42a16b23edd776c60a6d35f593f9d...  1973584935   
59087  c68f4b96d444ba171104c82af97abd3804ec3c7f0b7d3f...  1792768375   

             account  challenges         twitch              ladder  \
3724   Mamukuiskaaja          26            NaN  Hardcore Harbinger   
4406     kenntanaka1          22            NaN       SSF Harbinger   
36748        telmock          31        telmock       SSF Harbinger   
41356           MccM          21            NaN  Hardcore Harbinger   
59087     MushyMushy          33  actuallymushi  Hardcore Harbinger   

       special_class  stat_Str  stat_Int  stat_Dex  stats_num  
3724               1         1         0         1          2  
4406               1         0         1         0          1  
36748              1         0         0         1          1  
41356              1         1         1         0          2  
59087              1         1         0         0          1

#### plus...

클래스들의 특성에 따라 광딜, 도트딜, 강화약화주력 이런식으로 나눠볼까 했는데 역시 게임을 직접해보지 않고서야 아무리 tier들을 읽어보아도(https://gamerant.com/best-path-of-exile-ascendancy-classes-ranked/) 이해할 수 없으므로 이렇게만 나누고 분석을 시작해보고자 한다.

이후 알게되면서 필요하다면 추가 라벨링을 통해 분석을 할 예정이다.